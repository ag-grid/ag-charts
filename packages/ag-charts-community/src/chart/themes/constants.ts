import type {
    AgCartesianAxisPosition,
    AgCartesianAxisType,
    AgChartLegendPosition,
    AgPolarAxisShape,
    AgPolarAxisType,
    FontWeight,
} from '../../options/agChartOptions';

export const FONT_SIZE = {
    SMALL: 12,
    MEDIUM: 13,
    LARGE: 17,
};

export const BOLD: FontWeight = 'bold';
export const NORMAL: FontWeight = 'normal';

export const BOTTOM: AgChartLegendPosition = 'bottom';
const LEFT: AgCartesianAxisPosition = 'left';
const TOP: AgCartesianAxisPosition = 'top';
export const CARTESIAN_AXIS_POSITIONS = {
    BOTTOM,
    LEFT,
    TOP,
};

const CATEGORY: AgCartesianAxisType = 'category';
const NUMBER: AgCartesianAxisType = 'number';
const TIME: AgCartesianAxisType = 'time';
const LOG: AgCartesianAxisType = 'log';
export const CARTESIAN_AXIS_TYPES = {
    CATEGORY,
    NUMBER,
    TIME,
    LOG,
};

const ANGLE_CATEGORY: AgPolarAxisType = 'angle-category';
const ANGLE_NUMBER: AgPolarAxisType = 'angle-number';
const RADIUS_CATEGORY: AgPolarAxisType = 'radius-category';
const RADIUS_NUMBER: AgPolarAxisType = 'radius-number';
export const POLAR_AXIS_TYPES = {
    ANGLE_CATEGORY,
    ANGLE_NUMBER,
    RADIUS_CATEGORY,
    RADIUS_NUMBER,
};

export const CIRCLE: AgPolarAxisShape = 'circle';
