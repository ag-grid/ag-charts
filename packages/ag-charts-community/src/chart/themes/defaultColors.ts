export const DEFAULT_FILLS = {
    BLUE: '#5090dc',
    ORANGE: '#ffa03a',
    GREEN: '#459d55',
    CYAN: '#34bfe1',
    YELLOW: '#e1cc00',
    VIOLET: '#9669cb',
    GRAY: '#b5b5b5',
    MAGENTA: '#bd5aa7',
    BROWN: '#8a6224',
    RED: '#ef5452',
};

export const DEFAULT_STROKES = {
    BLUE: '#2b5c95',
    ORANGE: '#cc6f10',
    GREEN: '#1e652e',
    CYAN: '#18859e',
    YELLOW: '#a69400',
    VIOLET: '#603c88',
    GRAY: '#575757',
    MAGENTA: '#7d2f6d',
    BROWN: '#4f3508',
    RED: '#a82529',
};

export type DefaultColors = { fills: { [key: string]: string }; strokes: { [key: string]: string } };
