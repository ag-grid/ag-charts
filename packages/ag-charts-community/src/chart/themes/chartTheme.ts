import type {
    AgChartTheme,
    AgChartThemeOptions,
    AgChartThemeOverrides,
    AgChartThemePalette,
    AgCommonThemeableChartOptions,
    InteractionRange,
} from '../../options/agChartOptions';
import { jsonMerge, jsonWalk } from '../../util/json';
import { deepMerge } from '../../util/object';
import { AXIS_TYPES, getAxisThemeTemplate } from '../factory/axisTypes';
import { CHART_TYPES, type ChartType, getChartDefaults } from '../factory/chartTypes';
import { getLegendThemeTemplates } from '../factory/legendTypes';
import { getSeriesThemeTemplate } from '../factory/seriesTypes';
import { BOTTOM, FONT_SIZE, NORMAL } from './constants';
import { DEFAULT_FILLS, DEFAULT_STROKES } from './defaultColors';
import {
    DEFAULT_AXIS_GRID_COLOUR,
    DEFAULT_AXIS_LINE_COLOUR,
    DEFAULT_BACKGROUND_COLOUR,
    DEFAULT_CROSS_LINES_COLOUR,
    DEFAULT_DIVERGING_SERIES_COLOUR_RANGE,
    DEFAULT_FONT_FAMILY,
    DEFAULT_INSIDE_SERIES_LABEL_COLOUR,
    DEFAULT_INVERTED_LABEL_COLOUR,
    DEFAULT_LABEL_COLOUR,
    DEFAULT_MUTED_LABEL_COLOUR,
    DEFAULT_POLAR_SERIES_STROKE,
    DEFAULT_SHADOW_COLOUR,
    DEFAULT_WATERFALL_SERIES_CONNECTOR_LINE_STROKE,
    DEFAULT_WATERFALL_SERIES_NEGATIVE_COLOURS,
    DEFAULT_WATERFALL_SERIES_POSITIVE_COLOURS,
    DEFAULT_WATERFALL_SERIES_TOTAL_COLOURS,
    EXTENDS_AXES_DEFAULTS,
    EXTENDS_AXES_GRID_LINE_DEFAULTS,
    EXTENDS_AXES_LABEL_DEFAULTS,
    EXTENDS_AXES_LINE_DEFAULTS,
    EXTENDS_AXES_TICK_DEFAULTS,
    EXTENDS_CARTESIAN_MARKER_DEFAULTS,
    EXTENDS_CHART_DEFAULTS,
    EXTENDS_LEGEND_DEFAULTS,
    EXTENDS_LEGEND_ITEM_DEFAULTS,
    EXTENDS_LEGEND_ITEM_MARKER_DEFAULTS,
    EXTENDS_SERIES_DEFAULTS,
    OVERRIDE_SERIES_LABEL_DEFAULTS,
} from './symbols';

const DEFAULT_BACKGROUND_FILL = 'white';

const palette: AgChartThemePalette = {
    fills: Array.from(Object.values(DEFAULT_FILLS)),
    strokes: Array.from(Object.values(DEFAULT_STROKES)),
};

type ChartTypeConfig = {
    seriesTypes: string[];
    commonOptions: (keyof AgCommonThemeableChartOptions)[];
};
const CHART_TYPE_CONFIG: { [k in ChartType]: ChartTypeConfig } = {
    get cartesian(): ChartTypeConfig {
        return { seriesTypes: CHART_TYPES.cartesianTypes, commonOptions: ['zoom', 'navigator'] };
    },
    get polar(): ChartTypeConfig {
        return { seriesTypes: CHART_TYPES.polarTypes, commonOptions: [] };
    },
    get hierarchy(): ChartTypeConfig {
        return { seriesTypes: CHART_TYPES.hierarchyTypes, commonOptions: [] };
    },
};
const CHART_TYPE_SPECIFIC_COMMON_OPTIONS = Object.values(CHART_TYPE_CONFIG).reduce<
    (keyof AgCommonThemeableChartOptions)[]
>((r, { commonOptions }) => [...r, ...commonOptions], []);

export class ChartTheme {
    readonly palette: AgChartThemePalette;

    protected getPalette(): AgChartThemePalette {
        return palette;
    }

    readonly config: any;

    private static getAxisDefaults() {
        return {
            top: {},
            right: {},
            bottom: {},
            left: {},
            title: {
                enabled: false,
                text: 'Axis Title',
                spacing: 25,
                fontStyle: undefined,
                fontWeight: NORMAL,
                fontSize: FONT_SIZE.MEDIUM,
                fontFamily: DEFAULT_FONT_FAMILY,
                color: DEFAULT_LABEL_COLOUR,
            },
            label: {
                fontStyle: undefined,
                fontWeight: undefined,
                fontSize: FONT_SIZE.SMALL,
                fontFamily: DEFAULT_FONT_FAMILY,
                padding: 5,
                rotation: undefined,
                color: DEFAULT_LABEL_COLOUR,
                formatter: undefined,
                avoidCollisions: true,
            },
            line: {
                enabled: true,
                width: 1,
                color: DEFAULT_AXIS_LINE_COLOUR,
            },
            tick: {
                enabled: false,
                width: 1,
                color: DEFAULT_AXIS_LINE_COLOUR,
            },
            gridLine: {
                enabled: true,
                style: [
                    {
                        stroke: DEFAULT_AXIS_GRID_COLOUR,
                        lineDash: [],
                    },
                ],
            },
            crossLines: {
                enabled: false,
                fill: DEFAULT_CROSS_LINES_COLOUR,
                stroke: DEFAULT_CROSS_LINES_COLOUR,
                fillOpacity: 0.1,
                strokeWidth: 1,
                label: {
                    enabled: false,
                    fontStyle: undefined,
                    fontWeight: undefined,
                    fontSize: FONT_SIZE.SMALL,
                    fontFamily: DEFAULT_FONT_FAMILY,
                    padding: 5,
                    color: DEFAULT_LABEL_COLOUR,
                },
            },
        };
    }

    private static getSeriesDefaults() {
        return {
            tooltip: {
                enabled: true,
                renderer: undefined,
            },
            visible: true,
            showInLegend: true,
            highlightStyle: {
                item: {
                    fill: 'white',
                    fillOpacity: 0.33,
                    stroke: `rgba(0, 0, 0, 0.4)`,
                    strokeWidth: 2,
                },
                series: {
                    dimOpacity: 1,
                },
                text: {
                    color: 'black',
                },
            },
            nodeClickRange: 'exact' as InteractionRange,
        };
    }

    private static getCartesianSeriesMarkerDefaults() {
        return {
            enabled: true,
            shape: 'circle',
            size: 7,
            strokeWidth: 1,
            formatter: undefined,
        };
    }

    private static getLegendItemMarkerDefaults() {
        return {
            shape: undefined,
            size: 15,
            padding: 8,
        };
    }

    private static getCaptionWrappingDefaults() {
        return 'hyphenate' as const;
    }

    private static getChartDefaults() {
        return {
            background: {
                visible: true,
                fill: DEFAULT_BACKGROUND_COLOUR,
            },
            padding: {
                top: 20,
                right: 20,
                bottom: 20,
                left: 20,
            },
            title: {
                enabled: false,
                text: 'Title',
                fontStyle: undefined,
                fontWeight: NORMAL,
                fontSize: FONT_SIZE.LARGE,
                fontFamily: DEFAULT_FONT_FAMILY,
                color: DEFAULT_LABEL_COLOUR,
                wrapping: ChartTheme.getCaptionWrappingDefaults(),
            },
            subtitle: {
                enabled: false,
                text: 'Subtitle',
                spacing: 20,
                fontStyle: undefined,
                fontWeight: undefined,
                fontSize: FONT_SIZE.MEDIUM,
                fontFamily: DEFAULT_FONT_FAMILY,
                color: DEFAULT_MUTED_LABEL_COLOUR,
                wrapping: ChartTheme.getCaptionWrappingDefaults(),
            },
            footnote: {
                enabled: false,
                text: 'Footnote',
                spacing: 20,
                fontStyle: undefined,
                fontWeight: undefined,
                fontSize: FONT_SIZE.MEDIUM,
                fontFamily: DEFAULT_FONT_FAMILY,
                color: 'rgb(140, 140, 140)',
                wrapping: ChartTheme.getCaptionWrappingDefaults(),
            },
            legend: {
                position: BOTTOM,
                spacing: 30,
                listeners: {},
                item: {
                    paddingX: 16,
                    paddingY: 8,
                    marker: ChartTheme.getLegendItemMarkerDefaults(),
                    label: {
                        color: DEFAULT_LABEL_COLOUR,
                        fontStyle: undefined,
                        fontWeight: undefined,
                        fontSize: FONT_SIZE.SMALL,
                        fontFamily: DEFAULT_FONT_FAMILY,
                        formatter: undefined,
                    },
                },
                reverseOrder: false,
                pagination: {
                    marker: {
                        size: 12,
                    },
                    activeStyle: {
                        fill: DEFAULT_LABEL_COLOUR,
                    },
                    inactiveStyle: {
                        fill: DEFAULT_MUTED_LABEL_COLOUR,
                    },
                    highlightStyle: {
                        fill: DEFAULT_LABEL_COLOUR,
                    },
                    label: {
                        color: DEFAULT_LABEL_COLOUR,
                    },
                },
            },
            tooltip: {
                enabled: true,
                range: 'nearest' as InteractionRange,
                delay: 0,
            },
            listeners: {},
        };
    }

    private static readonly cartesianAxisDefault = {
        number: {
            ...ChartTheme.getAxisDefaults(),
            line: {
                ...ChartTheme.getAxisDefaults().line,
                // enabled: false, TODO: fix breaking tests when this is set, use width: 0 for now
                width: 0,
            },
        },
        log: {
            ...ChartTheme.getAxisDefaults(),
            base: 10,
            line: {
                ...ChartTheme.getAxisDefaults().line,
                // enabled: false, TODO: fix breaking tests when this is set, use width: 0 for now
                width: 0,
            },
        },
        category: {
            ...ChartTheme.getAxisDefaults(),
            groupPaddingInner: 0.1,
            label: {
                ...ChartTheme.getAxisDefaults().label,
                autoRotate: true,
            },
            gridLine: {
                ...ChartTheme.getAxisDefaults().gridLine,
                enabled: false,
            },
        },
        'grouped-category': {
            ...ChartTheme.getAxisDefaults(),
        },
        time: {
            ...ChartTheme.getAxisDefaults(),
            gridLine: {
                ...ChartTheme.getAxisDefaults().gridLine,
                enabled: false,
            },
        },
    };

    constructor(options?: AgChartTheme) {
        options = deepMerge({}, options ?? {}) as AgChartThemeOptions;
        const { overrides = null, palette = null } = options;

        const defaults = this.createChartConfigPerChartType(this.getDefaults());

        if (overrides) {
            const { common } = overrides;

            const applyOverrides = <K extends keyof typeof defaults>(
                seriesTypes: string[],
                overrideOpts: AgChartThemeOverrides[K]
            ) => {
                if (!overrideOpts) return;
                for (const s of seriesTypes) {
                    const seriesType = s as keyof AgChartThemeOverrides;
                    defaults[seriesType] = deepMerge(defaults[seriesType], overrideOpts);
                }
            };
            for (const [, { seriesTypes, commonOptions }] of Object.entries(CHART_TYPE_CONFIG)) {
                const cleanedCommon = { ...common };
                for (const commonKey of CHART_TYPE_SPECIFIC_COMMON_OPTIONS) {
                    if (!commonOptions.includes(commonKey)) {
                        delete cleanedCommon[commonKey];
                    }
                }
                applyOverrides(seriesTypes, cleanedCommon);
            }

            CHART_TYPES.seriesTypes.forEach((s) => {
                const seriesType = s as keyof AgChartThemeOverrides;
                if (overrides[seriesType]) {
                    defaults[seriesType] = deepMerge(defaults[seriesType], overrides[seriesType]);
                }
            });
        }
        this.palette = palette ?? this.getPalette();

        this.config = Object.freeze(this.templateTheme(defaults));
    }

    private createChartConfigPerChartType(config: AgChartThemeOverrides) {
        Object.entries(CHART_TYPE_CONFIG).forEach(([nextType, { seriesTypes }]) => {
            const typeDefaults = getChartDefaults(nextType as ChartType) as any;

            seriesTypes.forEach((next) => {
                const alias = next as keyof AgChartThemeOverrides;
                if (!config[alias]) {
                    config[alias] = {};
                    deepMerge(config[alias], typeDefaults);
                }
            });
        });

        return config;
    }

    private getDefaults(): AgChartThemeOverrides {
        let defaults = {};

        const getChartTypeDefaults = (chartType: ChartType) => {
            return {
                ...getLegendThemeTemplates(),
                ...ChartTheme.getChartDefaults(),
                ...getChartDefaults(chartType),
            };
        };

        const getOverridesByType = (chartType: ChartType, seriesTypes: string[]) => {
            const chartDefaults = getChartTypeDefaults(chartType) as any;
            const result: Record<string, { series?: {}; axes?: {} }> = {};
            for (const seriesType of seriesTypes) {
                result[seriesType] ??= deepMerge({}, chartDefaults);
                const axes: Record<string, {}> = (result[seriesType].axes ??= {});

                const template = getSeriesThemeTemplate(seriesType);
                if (template) {
                    result[seriesType].series = deepMerge(result[seriesType].series, template);
                }

                for (const axisType of AXIS_TYPES.axesTypes) {
                    const template = getAxisThemeTemplate(axisType);
                    if (chartType === 'cartesian') {
                        axes[axisType] = deepMerge(
                            axes[axisType],
                            (ChartTheme.cartesianAxisDefault as any)[axisType] ?? {}
                        );
                    }
                    if (template) {
                        axes[axisType] = deepMerge(axes[axisType], template);
                    }
                }
            }

            return result;
        };

        defaults = deepMerge(defaults, getOverridesByType('cartesian', CHART_TYPES.cartesianTypes));
        defaults = deepMerge(defaults, getOverridesByType('polar', CHART_TYPES.polarTypes));
        defaults = deepMerge(defaults, getOverridesByType('hierarchy', CHART_TYPES.hierarchyTypes));

        return defaults;
    }

    templateTheme<T>(themeTemplate: T): T {
        const themeInstance = jsonMerge([themeTemplate]);
        const { extensions, properties } = this.getTemplateParameters();

        jsonWalk(themeInstance, (_, node) => {
            if (node['__extends__']) {
                const key = node['__extends__'];
                const source = extensions.get(key);
                if (source == null) {
                    throw new Error('AG Charts - no template variable provided for: ' + key);
                }
                Object.keys(source).forEach((key) => {
                    if (!(key in node)) {
                        node[key] = source[key];
                    }
                });
                delete node['__extends__'];
            }
            if (node['__overrides__']) {
                const key = node['__overrides__'];
                const source = extensions.get(key);
                if (source == null) {
                    throw new Error('AG Charts - no template variable provided for: ' + key);
                }
                Object.assign(node, source);
                delete node['__overrides__'];
            }

            if (Array.isArray(node)) {
                for (let i = 0; i < node.length; i++) {
                    const symbol = node[i];
                    if (properties.has(symbol)) {
                        node[i] = properties.get(symbol);
                    }
                }
            } else {
                for (const [name, value] of Object.entries(node)) {
                    if (properties.has(value)) {
                        node[name] = properties.get(value);
                    }
                }
            }
        });

        return themeInstance;
    }

    protected static getWaterfallSeriesDefaultPositiveColors() {
        return {
            fill: DEFAULT_FILLS.BLUE,
            stroke: DEFAULT_STROKES.BLUE,
        };
    }

    protected static getWaterfallSeriesDefaultNegativeColors() {
        return {
            fill: DEFAULT_FILLS.ORANGE,
            stroke: DEFAULT_STROKES.ORANGE,
        };
    }

    protected static getWaterfallSeriesDefaultTotalColors() {
        return {
            fill: DEFAULT_FILLS.GRAY,
            stroke: DEFAULT_STROKES.GRAY,
        };
    }

    getTemplateParameters() {
        const extensions = new Map();
        extensions.set(EXTENDS_CHART_DEFAULTS, ChartTheme.getChartDefaults());
        extensions.set(EXTENDS_AXES_DEFAULTS, ChartTheme.getAxisDefaults());
        extensions.set(EXTENDS_LEGEND_DEFAULTS, ChartTheme.getChartDefaults().legend);
        extensions.set(EXTENDS_LEGEND_ITEM_DEFAULTS, ChartTheme.getChartDefaults().legend.item);
        extensions.set(EXTENDS_LEGEND_ITEM_MARKER_DEFAULTS, ChartTheme.getLegendItemMarkerDefaults());
        extensions.set(EXTENDS_AXES_LABEL_DEFAULTS, ChartTheme.getAxisDefaults().label);
        extensions.set(EXTENDS_AXES_LINE_DEFAULTS, ChartTheme.getAxisDefaults().line);
        extensions.set(EXTENDS_AXES_TICK_DEFAULTS, ChartTheme.getAxisDefaults().tick);
        extensions.set(EXTENDS_AXES_GRID_LINE_DEFAULTS, ChartTheme.getAxisDefaults().gridLine);
        extensions.set(EXTENDS_SERIES_DEFAULTS, ChartTheme.getSeriesDefaults());
        extensions.set(OVERRIDE_SERIES_LABEL_DEFAULTS, {});
        extensions.set(EXTENDS_CARTESIAN_MARKER_DEFAULTS, ChartTheme.getCartesianSeriesMarkerDefaults());

        const properties = new Map();
        properties.set(DEFAULT_FONT_FAMILY, 'Verdana, sans-serif');
        properties.set(DEFAULT_LABEL_COLOUR, 'rgb(70, 70, 70)');
        properties.set(DEFAULT_INVERTED_LABEL_COLOUR, 'white');
        properties.set(DEFAULT_MUTED_LABEL_COLOUR, 'rgb(140, 140, 140)');
        properties.set(DEFAULT_AXIS_GRID_COLOUR, 'rgb(224,234,241)');
        properties.set(DEFAULT_AXIS_LINE_COLOUR, 'rgb(195, 195, 195)');
        properties.set(DEFAULT_CROSS_LINES_COLOUR, 'rgb(70, 70, 70)');
        properties.set(DEFAULT_INSIDE_SERIES_LABEL_COLOUR, DEFAULT_BACKGROUND_FILL);
        properties.set(DEFAULT_BACKGROUND_COLOUR, DEFAULT_BACKGROUND_FILL);
        properties.set(DEFAULT_SHADOW_COLOUR, 'rgba(0, 0, 0, 0.5)');
        properties.set(DEFAULT_DIVERGING_SERIES_COLOUR_RANGE, ['rgb(247,144,110)', 'rgb(140,236,121)']);
        properties.set(DEFAULT_POLAR_SERIES_STROKE, DEFAULT_BACKGROUND_FILL);
        properties.set(DEFAULT_WATERFALL_SERIES_POSITIVE_COLOURS, ChartTheme.getWaterfallSeriesDefaultPositiveColors());
        properties.set(DEFAULT_WATERFALL_SERIES_NEGATIVE_COLOURS, ChartTheme.getWaterfallSeriesDefaultNegativeColors());
        properties.set(DEFAULT_WATERFALL_SERIES_TOTAL_COLOURS, ChartTheme.getWaterfallSeriesDefaultTotalColors());
        properties.set(
            DEFAULT_WATERFALL_SERIES_CONNECTOR_LINE_STROKE,
            ChartTheme.getWaterfallSeriesDefaultTotalColors().stroke
        );

        return {
            extensions,
            properties,
        };
    }
}
