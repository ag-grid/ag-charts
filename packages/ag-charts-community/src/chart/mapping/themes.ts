import type { AgChartTheme, AgChartThemeName, AgChartThemeOverrides } from '../../options/agChartOptions';
import { jsonMerge } from '../../util/json';
import { ChartTheme } from '../themes/chartTheme';
import { DarkTheme } from '../themes/darkTheme';
import { MaterialDark } from '../themes/materialDark';
import { MaterialLight } from '../themes/materialLight';
import { PastelDark } from '../themes/pastelDark';
import { PastelLight } from '../themes/pastelLight';
import { SolarDark } from '../themes/solarDark';
import { SolarLight } from '../themes/solarLight';
import { VividDark } from '../themes/vividDark';
import { VividLight } from '../themes/vividLight';

type ThemeMap = { [key in AgChartThemeName | 'undefined' | 'null']?: () => ChartTheme };

const lightTheme = () => new ChartTheme();
const darkTheme = () => new DarkTheme();

const lightThemes: ThemeMap = {
    undefined: lightTheme,
    null: lightTheme,
    'ag-default': lightTheme,
    'ag-material': () => new MaterialLight(),
    'ag-pastel': () => new PastelLight(),
    'ag-solar': () => new SolarLight(),
    'ag-vivid': () => new VividLight(),
};

const darkThemes: ThemeMap = {
    undefined: darkTheme,
    null: darkTheme,
    'ag-default-dark': darkTheme,
    'ag-material-dark': () => new MaterialDark(),
    'ag-pastel-dark': () => new PastelDark(),
    'ag-solar-dark': () => new SolarDark(),
    'ag-vivid-dark': () => new VividDark(),
};

export const themes: ThemeMap = {
    ...darkThemes,
    ...lightThemes,
};

export function getChartTheme(value?: string | ChartTheme | AgChartTheme): ChartTheme {
    if (value instanceof ChartTheme) {
        return value;
    }

    const stockTheme = themes[value as AgChartThemeName];
    if (stockTheme) {
        return stockTheme();
    }

    value = value as AgChartTheme;

    // Flatten recursive themes.
    const overrides: AgChartThemeOverrides[] = [];
    let palette;
    while (typeof value === 'object') {
        overrides.push(value.overrides ?? {});

        // Use first palette found, they can't be merged.
        if (value.palette && palette == null) {
            palette = value.palette;
        }

        value = value.baseTheme;
    }
    overrides.reverse();

    const flattenedTheme = {
        baseTheme: value,
        overrides: jsonMerge(overrides),
        ...(palette ? { palette } : {}),
    };

    if (flattenedTheme.baseTheme || flattenedTheme.overrides) {
        const baseTheme: any = getChartTheme(flattenedTheme.baseTheme);
        return new baseTheme.constructor(flattenedTheme);
    }

    return lightTheme();
}
