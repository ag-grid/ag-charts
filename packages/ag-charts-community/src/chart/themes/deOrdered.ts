import type { AgChartThemePalette } from '../../options/agChartOptions';
import { ChartTheme } from './chartTheme';

const palette: AgChartThemePalette = {
    fills: [
        '#436ff4',
        '#f2a602',
        '#9a7bff',
        '#e9e201',
        '#d165d2',
        '#21b448',
        '#f0598b',
        '#00b9a2',
        '#f47348',
        '#00aee4',
    ],
    strokes: [
        '#132baf',
        '#a55f00',
        '#623bba',
        '#8f8500',
        '#8f2291',
        '#007500',
        '#a90352',
        '#007762',
        '#ae3200',
        '#006fa3',
    ],
};

export class DeOrdered extends ChartTheme {
    protected override getPalette(): AgChartThemePalette {
        return palette;
    }
}
