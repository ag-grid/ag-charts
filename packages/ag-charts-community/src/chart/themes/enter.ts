import type { AgChartThemePalette } from '../../options/agChartOptions';
import { ChartTheme } from './chartTheme';

const palette: AgChartThemePalette = {
    "fills": [
        "#4F81BD",
        "#F2903F",
        "#4D9859",
        "#43AEB5",
        "#D85451",
        "#89A949",
        "#8664AF",
        "#82602E",
        "#A55492",
        "#8E8E8E",
      ],
    "strokes": [
        '#23558d',
        '#bf6100',
        '#5e7b11',
        '#008087',
        '#a42026',
        '#1b6b2e',
        '#5b3980',
        '#563700',
        '#752865',
        '#626262',
      ]
};

export class Enter extends ChartTheme {
    protected override getPalette(): AgChartThemePalette {
        return palette;
    }
}
