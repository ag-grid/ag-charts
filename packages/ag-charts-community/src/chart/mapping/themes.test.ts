import { describe, expect, it } from '@jest/globals';

import { AgCharts } from '../../api/agCharts';
import type {
    AgBarSeriesOptions,
    AgChartOptions,
    AgChartTheme,
    AgChartThemeName,
    AgChartThemePalette,
} from '../../options/agChartOptions';
import type { Chart } from '../chart';
import {
    deproxy,
    expectWarningsCalls,
    prepareTestOptions,
    setupMockConsole,
    waitForChartStability,
} from '../test/utils';
import { themesMap } from './themes';

describe('themes module', () => {
    setupMockConsole();

    const getPalette = (themeName: AgChartThemeName): AgChartThemePalette | undefined => {
        const ctr = themesMap.get(themeName);
        if (ctr !== undefined) {
            return ctr().palette;
        }
    };

    const getActualPalette = (chart: Chart) => {
        let result = undefined;
        for (const series of deproxy(chart).processedOptions.series ?? []) {
            result ??= { fills: [] as string[], strokes: [] as string[] };

            expect(series.type).toEqual('bar');
            const barseries = series as AgBarSeriesOptions;
            if (barseries.fill !== undefined) {
                result.fills.push(barseries.fill);
            }
            if (barseries.stroke !== undefined) {
                result.strokes.push(barseries.stroke);
            }
        }
        return result;
    };

    const opts: AgChartOptions = prepareTestOptions({
        data: [
            { n: 'A', v0: 4.2, v1: 5.6, v2: 8.6, v3: 8.1, v4: 6.4, v5: 1.3, v6: 6.4, v7: 1.3, v8: 2.2, v9: 8.7 },
            { n: 'B', v0: 1.8, v1: 7.1, v2: 8.4, v3: 1.3, v4: 6.8, v5: 5.5, v6: 2.7, v7: 4.8, v8: 4.8, v9: 5.2 },
            { n: 'C', v0: 7.1, v1: 7.4, v2: 1.9, v3: 9.8, v4: 1.3, v5: 4.4, v6: 8.3, v7: 9.5, v8: 1.3, v9: 0.9 },
            { n: 'D', v0: 3.5, v1: 9.2, v2: 4.2, v3: 2.5, v4: 6.3, v5: 4.4, v6: 5.9, v7: 2.2, v8: 6.8, v9: 0.1 },
            { n: 'E', v0: 9.0, v1: 2.8, v2: 1.9, v3: 7.4, v4: 5.9, v5: 8.1, v6: 0.6, v7: 7.6, v8: 3.0, v9: 3.4 },
        ],
        series: [
            { type: 'bar', xKey: 'n', yKey: 'v0', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v1', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v2', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v3', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v4', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v5', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v6', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v7', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v8', stacked: true },
            { type: 'bar', xKey: 'n', yKey: 'v9', stacked: true },
        ],
    });

    it('should show 1 warning for invalid theme type', async () => {
        const chart = deproxy(
            AgCharts.create({
                ...opts,
                theme: true as unknown as AgChartTheme,
            })
        );
        await waitForChartStability(chart);

        expectWarningsCalls().toMatchInlineSnapshot(`
[
  [
    "AG Charts - Option theme cannot be set to [true]; expecting a keyword such as 'ag-default', 'ag-default-dark', 'ag-sheets', 'ag-sheets-dark', 'ag-polychroma', and 2 more or 'ag-material-dark' or an object, ignoring.",
  ],
  [
    "AG Charts - invalid theme value type boolean, expected object or string.",
  ],
]
`);
    });

    test('missing strokes', async () => {
        const chart = deproxy(
            AgCharts.create({
                ...opts,
                theme: {
                    baseTheme: 'ag-default-dark',
                    palette: {
                        fills: ['#5C2983', '#0076C5', '#21B372', '#FDDE02', '#F76700', '#D30018'],
                    } as AgChartThemePalette,
                },
            })
        );
        await waitForChartStability(chart);

        expect(getActualPalette(chart)?.strokes).toEqual(getPalette('ag-default-dark')?.strokes);
    });

    test('missing fills', async () => {
        const chart = deproxy(
            AgCharts.create({
                ...opts,
                theme: {
                    baseTheme: 'ag-default-dark',
                    palette: {
                        strokes: ['black'],
                    } as AgChartThemePalette,
                },
            })
        );
        await waitForChartStability(chart);

        expect(getActualPalette(chart)?.fills).toEqual(getPalette('ag-default-dark')?.fills);
    });

    it('should show 3 warnings for invalid types', async () => {
        const chart = deproxy(
            AgCharts.create({
                ...opts,
                theme: {
                    baseTheme: NaN,
                    palette: 'foobar',
                    overrides: true,
                } as unknown as AgChartTheme,
            })
        );
        await waitForChartStability(chart);

        expectWarningsCalls().toMatchInlineSnapshot(`
[
  [
    "AG Charts - Option theme.baseTheme cannot be set to [NaN]; expecting a keyword such as 'ag-default', 'ag-default-dark', 'ag-sheets', 'ag-sheets-dark', 'ag-polychroma', and 2 more or 'ag-material-dark' or an object, ignoring.",
  ],
  [
    "AG Charts - Option theme.overrides cannot be set to [true]; expecting an object, ignoring.",
  ],
  [
    "AG Charts - Option theme.palette cannot be set to ["foobar"]; expecting an object, ignoring.",
  ],
  [
    "AG Charts - invalid theme value type object, expected object or string.",
  ],
]
`);
    });

    it('should show 2 warnings for invalid types - palette', async () => {
        const chart = deproxy(
            AgCharts.create({
                ...opts,
                theme: {
                    baseTheme: 'ag-default-dark',
                    palette: {
                        fills: 'red',
                        strokes: 'black',
                    } as unknown as AgChartThemePalette,
                },
            })
        );
        await waitForChartStability(chart);

        expectWarningsCalls().toMatchInlineSnapshot(`
[
  [
    "AG Charts - Option theme.palette.fills cannot be set to ["red"]; expecting a string array, ignoring.",
  ],
  [
    "AG Charts - Option theme.palette.strokes cannot be set to ["black"]; expecting a string array, ignoring.",
  ],
  [
    "AG Charts - invalid theme value type object, expected object or string.",
  ],
]
`);
    });
});
