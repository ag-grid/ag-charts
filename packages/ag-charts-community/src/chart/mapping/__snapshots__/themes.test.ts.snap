// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`themes module should show 1 warning for invalid theme type 1`] = `
[
  [
    "AG Charts - invalid theme value type boolean, expected object or string.",
  ],
]
`;

exports[`themes module should show 2 warnings for invalid types - palette 1`] = `
[
  [
    "AG Charts - theme.overrides.fills must be undefined or an array",
  ],
  [
    "AG Charts - theme.overrides.strokes must be undefined or an array",
  ],
]
`;

exports[`themes module should show 3 warnings for invalid types 1`] = `
[
  [
    "AG Charts - invalid theme.baseTheme type number, expected (string | object).",
  ],
  [
    "AG Charts - invalid theme.overrides type boolean, expected object.",
  ],
  [
    "AG Charts - invalid theme.palette type string, expected object.",
  ],
]
`;
